tools/checklocks/state.go

```diff
@@ -346,30 +413,33 @@ func (l *lockState) String() string {
 
 	var locksHeld []string
 	for k, lockType := range l.mutexes {
-		if lockType == locked {
+		if lockType.isLocked() {
 			locksHeld = append(locksHeld, k)
 		}
 	}
-	// sort for comparing results 
+	// sort for comparing results
 	sort.Strings(locksHeld)
 
 	return strings.Join(locksHeld, ",")
 }
 
-// pushDefer pushes a defer onto the stack.
-func (l *lockState) pushDefer(d *ssa.Defer) {
-	l.modify()
-	l.defers = append(l.defers, d)
-}
-
-// popDefer pops a defer from the stack.
-func (l *lockState) popDefer() *ssa.Defer {
-	// Does not technically modify the underlying slice.
-	count := len(l.defers)
-	if count == 0 {
-		return nil
+// resolveDeferredUnlocks symbolically executes earlier `defer *.Unlock()` statements on
+// the lock state. This function is called in checkInstruction when a ssa.RunDefers is
+// encountered.
+func (l *lockState) resolveDeferredUnlocks() (string, bool) {
+	for k, lockType := range l.mutexes {
+		if lockType.hasDeferredUnlock() {
+			// lock `k` is already unlocked, but there is also a deferred unlock
+			// return the invalid mutex name.
+			return k, false
+		}
 	}
-	d := l.defers[count-1]
-	l.defers = l.defers[:count-1]
-	return d
+	for k, lockType := range l.mutexes {
+		if lockType.isLocked() && lockType.hasDeferredUnlock() {
+			l.modify()
+			l.mutexes[k] = unlocked
```

Comment:
This could be `delete(l.mutexes, k)` instead.

Reply:
Either/or. Delete is less clear but more efficient. This can be addressed by adding a comment.